src/main/io/gps.c

```diff
@@ -1578,9 +2069,53 @@ static bool UBLOX_parse_gps(void)
     uint32_t i;
 
     *gpsPacketLogChar = LOG_IGNORED;
+#define CLSMSG(cls, msg) (((cls) << 8) | (msg))
+    switch (CLSMSG(_class, _msg_id)) {
+    case CLSMSG(CLASS_MON, MSG_MON_VER): {
+        char *token = strtok(_buffer.ver.swVersion, " ");
+        while (token != NULL) {
+            float f = fastA2F(token);
+            if (f > 0.0f) {
+                int major = strtol(token, NULL, 10);
+                int minor = strtol(&token[major > 9 ? 3 : 2], NULL, 10);
+                gpsData.monVer.swVersion.firmwareVersion.major = major;
+                gpsData.monVer.swVersion.firmwareVersion.minor = minor;
+                break;
+            }
+            token = strtok(NULL, " ");
+        }
 
-    switch (_msg_id) {
-    case MSG_POSLLH:
+#ifdef USE_GPS_DEBUG
+        // get software protocol version details
+        size_t j = 0;
+        while (j < 10) {
+            if (_buffer.ver.extension[j][0] == '\0') {
+                break;
+            }
+            token = strtok(&_buffer.ver.extension[j++][0], "=");
+            while (token != NULL) {
+                if (strcmp(token, "PROTVER") == 0) {
+                    token = strtok(NULL, "=");
+                    int major = strtol(token, NULL, 10);
+                    int minor = strtol(&token[major > 9 ? 3 : 2], NULL, 10);
+                    gpsData.monVer.swVersion.protocolVersion.major = major;
+                    gpsData.monVer.swVersion.protocolVersion.minor = minor;
+                    break;
+                }
+                token = strtok(NULL, "=");
+            }
+        }
+#endif
+
+        gpsData.monVer.hwVersion = strtoul(_buffer.ver.hwVersion, NULL, 16);
+        gpsData.unitVersion = ubloxParseVersion(gpsData.monVer.hwVersion);
+        gpsData.ubloxM7orAbove = gpsData.unitVersion >= UBX_VERSION_M7;
+        gpsData.ubloxM8orAbove = gpsData.unitVersion >= UBX_VERSION_M8;
+        gpsData.ubloxM9orAbove = gpsData.unitVersion >= UBX_VERSION_M9;
```

Comment:
```suggestion
        gpsData.ubloxM7orAbove = gpsData.unitVersion >= UBX_VERSION_M7;
        gpsData.ubloxM8orAbove = gpsData.unitVersion >= UBX_VERSION_M8;
        gpsData.ubloxM9orAbove = gpsData.unitVersion >= UBX_VERSION_M9;
```

These flags should be replaced with switch function for readability.

Reply:
What's hard to read about the way it is?

Reply:
Example:

![image](https://github.com/betaflight/betaflight/assets/8344830/dab316f9-9b20-475b-9812-0fd8693a3399)

![image](https://github.com/betaflight/betaflight/assets/8344830/c384cd5a-7d4a-4edd-b4e6-fb007bca7a63)

Reply:
@haslinghuis: To consider:
- This will fail when new version is added, but not using `default:`  will generate warning in that case (unused enum value). Also, it will clutter code when too much versions are necessary.
- `switch` will create new block for `break`, possibly making (this stage of) refactoring more complicated.

Reply:
Eventually it might be easier to have code separated in version modules?
Maybe letting user[s] define UBX version at compile time we can use macro's to save memory.
Would be interesting to follow as learning experience how we solve this.

Reply:
IMO bad idea in this case. Code is almost identical and difference is (mostly?) initialization only. This whole PR is mainly about not bothering user with UBX version.

And BF was taking long way to make targets universal. Now it is doing a bit in opposite direction, but I wouldn't push it too much

Reply:
> Eventually it might be easier to have code separated in version modules? Maybe letting user[s] define UBX version at compile time we can use macro's to save memory. Would be interesting to follow as learning experience how we solve this.

We have a few dimensions of variation we need to design into the next version of all this:
- There's a  big variation around the protocol being used (NMEA vs UBX). But in other ways, the state machine for the GPS module goes through similar states for both (they both get initialized, have normal operation states, etc.). We have a boatload of duplication across NMEA and UBX.
- The implementation of the states varies by platform/protocol version. Mostly it's the specific commands being sent to the module, but sometimes the logic varies as well.
- The number and sometimes sequence of states will vary, usually based on version but not always. I guess this depends on how the north-star state machine is designed.

The current code design does not clearly separate out the layers of the driver. This is a standard UART sensor driver, so we could use industry standard patterns for the module state vs. the protocol state vs. the transport (UART) state layers.

In the past, I've done some state machines with the states and their implementation in functions that are assembled using function pointers. It makes it really flexible to handle all the variation, but often more challenging to read and follow. Patterns for handling all that variation cleanly will be important for us.

Reply:
Can we agree to resolve this conversation for now?  We can propose a refactoring PR later?